solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,377 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+
+pragma solidity 0.8.17;
+
+import "../bridge/Bridge.sol";
+import "../bridge/Deposit.sol";
+
+/// @title TBTC Optimistic Minting
+/// @notice The Optimistic Minting mechanism allows to mint TBTC before
+///         TBTCVault receives the Bank balance. There are two permissioned sets
+///         in the system: Minters and Guardians, both set up in 1-of-n mode.
+///         Minters observe the revealed deposits and request minting TBTC.
+///         Any single Minter can perform this action. There is a 3 hours delay
+///         between the time of the request from a Minter to the time TBTC is
+///         minted. During the time of the delay, any Guardian can cancel the
+///         minting.
+/// @dev This functionality is a part of TBTCVault. It is implemented in
+///      a separate abstract contract to achieve better separation of concerns
+///      and easier-to-follow code.
+abstract contract TBTCOptimisticMinting is Ownable {
+    /// @notice The time that needs to pass between the moment the optimistic
+    ///         minting is requested and the moment optimistic minting is
+    ///         finalized with minting TBTC.
+    uint256 public constant OPTIMISTIC_MINTING_DELAY = 3 hours;
+
+    Bridge public bridge;
+
+    /// @notice Indicates if the given address is a Minter. Only Minters can
+    ///         request optimistic minting.
+    mapping(address => bool) public isMinter;
+
+    /// @notice Indicates if the given address is a Guardian. Only Guardians can
+    ///         cancel requested optimistic minting.
+    mapping(address => bool) public isGuardian;
+
+    /// @notice Indicates if the optimistic minting has been paused. Only the
+    ///         Governance can pause optimistic minting. Note that the pause of
+    ///         the optimistic minting does not stop the standard minting flow
+    ///         where wallets sweep deposits.
+    bool public isOptimisticMintingPaused;
```

Comment:
What do you think if we change this to `bool public isOptimisticMintingAllowed = true`? This way we would use fewer negations in the code which makes reading easier. E.g
Instead of 
```
modifier whenOptimisticMintingNotPaused() {
        require(!isOptimisticMintingPaused, "Optimistic minting paused");
        _;
    }
```
we can write
```
 modifier whenOptimisticMintingAllowed() {
        require(isOptimisticMintingAllowed, "Optimistic minting paused"); // paused -> disallowed?
        _;
    }
```
and use `whenOptimisticMintingAllowed` where applicable.

Reply:
I kind of prefer the term "paused" instead of "allowed", as this path is something extraordinary, and pausing minting is exactly what it is meant to do.
What is more, the variable will be by default initialized with `false`, so no need to initiate it at the contract deployment.

Reply:
+1 to @nkuba's answer. Additionally, I'd like to note that the current solution is similar to how OpenZeppelin [handles it](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/Pausable.sol).

Reply:
ok